spec/cancan/model_adapters/active_record_adapter_spec.rb

```diff
@@ -443,6 +443,199 @@ class User < ActiveRecord::Base
     expect(Article.accessible_by(ability)).to eq([article])
   end
 
+  it 'allows an empty array to be used as a condition for a has_many, but this is never a passing condition' do
+    a1 = Article.create!
+    a2 = Article.create!
+    a2.comments = [Comment.create!]
+
+    @ability.can :read, Article, comment_ids: []
+
+    expect(@ability.can?(:read, a1)).to eq(false)
+    expect(@ability.can?(:read, a2)).to eq(false)
+
+    expect(Article.accessible_by(@ability)).to eq([])
+
+    if CanCan::ModelAdapters::ActiveRecordAdapter.version_greater_or_equal?('5.0.0')
+      expect(@ability.model_adapter(Article, :read)).to generate_sql(%(
+  SELECT "articles".*
+  FROM "articles"
+  WHERE 1=0))
+    end
+  end
+
+  it 'allows a nil to be used as a condition for a has_many - with join' do
+    a1 = Article.create!
+    a2 = Article.create!
+    a2.comments = [Comment.create!]
+
+    @ability.can :read, Article, comments: { id: nil }
```

Comment:
would `can :read, Article, comments: nil` work as well? I'd expect so from 3.2.0.

And what about `can :read, Article, comments: []`? Maybe I'd expect this to work as well as `can :read, Article, comments: { id: [] }`

Reply:
Not currently. The cancan conditions need to match conditions you'd pass into AR#where. Both of your examples wouldn't work as regular queries.

```ruby
(byebug) Article.left_outer_joins(:comments).where(comments: nil)
#<ActiveRecord::Relation []>
(byebug) Article.left_outer_joins(:comments).where(comments: nil).to_sql
"SELECT \"articles\".* FROM \"articles\" LEFT OUTER JOIN \"comments\" ON \"comments\".\"article_id\" = \"articles\".\"id\" WHERE \"articles\".\"id\" IS NULL" # note how it looks for article.ids IS NULL, not comments.id
```

```ruby
(byebug) Article.left_outer_joins(:comments).where(comments: [])
#<ActiveRecord::Relation []>
(byebug) Article.left_outer_joins(:comments).where(comments: []).to_sql
"SELECT \"articles\".* FROM \"articles\" LEFT OUTER JOIN \"comments\" ON \"comments\".\"article_id\" = \"articles\".\"id\" WHERE 1=0"
```

Whereas with my example:

```ruby
(byebug) Article.left_outer_joins(:comments).where(comments: { id: nil })
#<ActiveRecord::Relation [#<Article id: 1, name: nil, created_at: "2020-12-14 00:50:15", updated_at: "2020-12-14 00:50:15", published: nil, secret: nil, priority: nil, category_id: nil, project_id: nil, user_id: nil>]>
(byebug) Article.left_outer_joins(:comments).where(comments: { id: nil }).to_sql
"SELECT \"articles\".* FROM \"articles\" LEFT OUTER JOIN \"comments\" ON \"comments\".\"article_id\" = \"articles\".\"id\" WHERE \"comments\".\"id\" IS NULL"
```

Reply:
I'm open to trying to support more use cases, but I think it'd add a bit more complexity as we'd need to translate specific values (`nils`, `[]`, etc) into the corresponding AR query statements.